### PR TITLE
[misc] Update `README.md`

### PR DESCRIPTION
**Pull request by yuanming-hu**
Related issue = #740 

[[**Preview**]](https://github.com/yuanming-hu/taichi/blob/readme2/README.md)

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

### CODE REVIEW

**Comment by yuanming-hu** [comment]
Now we have a gallery: https://github.com/yuanming-hu/taichi/blob/readme2/README.md#gallery

We have 3 slots left.
 - @k-ye could you please generate a 192x192 gif of pbf2d? 
 - @archibate could you help with `game-of-life` and `nbody_oscillator`?

After https://github.com/taichi-dev/taichi/pull/783 is reviewed and merged, `ti video [frame rate]` will gather all images under the current folder into an `mp4` video.

https://ezgif.com/ is a great online tool to convert `mp4` to `gif and for compressing your gifs.

**Comment by archibate** [comment]
Great! I'll polish nbodyosc before making its gif.
Btw, why no `ti gif` to directly generate gif?

**Comment by yuanming-hu** [comment]
> Great! I'll polish nbodyosc before making its gif.
> Btw, why no `ti gif` to directly generate gif?

Sounds good, thanks. We do have `ti gif X.mp4` to convert a video to gif. Maybe we can let `ti gif` directly generate gif when it's called without any arguments.

**Comment by xumingkuan** [comment]
I wonder if fractal.py is related to spatially sparse computation? Will we use fewer threads to paint the white pixels and spend more computation resources on black pixels?

**Comment by yuanming-hu** [comment]
> I wonder if fractal.py is related to spatially sparse computation? Will we use fewer threads to paint the white pixels and spend more computation resources on black pixels?

This sounds like a chicken-and-egg problem - if we don't do the computation, it's hard to know which part is white and which part is black :-) I would say `fractal` is not really suitable for sparse computing unless there's a way to quick skip inactive parts.

**Comment by xumingkuan** [comment]
So is there a better example to show "spatially sparse computation and differentiable programming"? Although `fractal` looks nice, people might get confused by seeing such an example following the above statement about sparse computing.

**Comment by k-ye** [comment]
Here you go :)

pbf2d

File: README.md [hunk 1]
```diff
@@ -3,7 +3,20 @@
   <h3> <a href="https://taichi.readthedocs.io/en/latest/"> Documentation </a> | <a href="https://taichi.readthedocs.io/en/latest/hello.html"> Tutorial </a> | <a href="https://github.com/yuanming-hu/difftaichi"> DiffTaichi </a> | <a href="https://github.com/yuanming-hu/taichi/tree/master/examples"> Examples </a> | <a href="https://taichi.readthedocs.io/en/latest/contributor_guide.html"> Contributor Guidelines </a> | <a href="https://forum.taichi.graphics/"> Forum </a> </h3>
 </div>
 
-### Installation [![Downloads](https://pepy.tech/badge/taichi-nightly)](https://pepy.tech/project/taichi-nightly)
+## Overview
+
+Taichi (太极) is a high-performance programming language and compiler, designed for computer graphics applications. It has native support for spatially sparse computation and differentiable programming.
+
+<img src="https://github.com/yuanming-hu/public_files/raw/master/graphics/taichi/fractal_code.png" height="270px"> <img src="https://raw.githubusercontent.com/yuanming-hu/public_files/master/graphics/taichi/fractal.gif" height="270px">
```
Comment (k-ye) [inline README.md:10]: Is it possible to directly embed the code with this link..?

https://github.com/taichi-dev/taichi/blob/db8d2d6eb8c4b09a8405449086f4e9f5858e7f0f/examples/fractal.py#L1-L31

Comment (yuanming-hu) [reply README.md:10]: Good idea. I just now tried that and it seems to take a huge space:

Screenshot from 2020-04-14 19-54-51


How about adding a link to the code image to let it point to the actual file?

Comment (k-ye) [reply README.md:10]: Sure :)

**Comment by archibate** [comment]
> So is there a better example to show "spatially sparse computation and differentiable programming"?

I thinks game_of_life can be a good idea to show off the usage of bitmasked. Only those active cells needs to be iterated. And we can also show them the speed boost by comparing the profiling result.

**Comment by yuanming-hu** [comment]
> > So is there a better example to show "spatially sparse computation and differentiable programming"?
> 
> I thinks game_of_life can be a good idea to show off the usage of bitmasked. Only those active cells needs to be iterated. And we can also show them the speed boost by comparing the profiling result.

My feeling is that both sparse and differentiable programming are advanced topics. We should keep the code sample as simple as possible, since most users may just want a system that allows them to write computer graphics code in Python. I'll think about the text organization here.

**Comment by archibate** [comment]
> allows them to write computer graphics code in Python

So taichi is the only option to write gpgpu shader code in py?

**Comment by k-ye** [comment]
> So taichi is the only option to write gpgpu shader code in py?

I think there are lots of Py wrappers around CUDA/OpenGL. Yet Taichi is a innovative way to hide all these complexities and unify them. And it’s not shader lang, but (a subset of) native python code, right :)?

IMHO the purpose of the REAMDE is to attract people, instead of trying to dump all the great features to them all at once.. That said, maybe AD deserves another section and/or a small example to illustrate what Taichi is capable of.

**Comment by yuanming-hu** [comment]
> IMHO the purpose of the REAMDE is to attract people, instead of trying to dump all the great features to them all at once.. That said, maybe AD deserves another section and/or a small example to illustrate what Taichi is capable of.

Agreed. It's always good to start easy. For this PR let's just include gifs of mostly basic features, as well as a brief introduction to the developer team. I'll update on advanced topics later in a different PR.

**Comment by yuanming-hu** [comment]
@archibate I added `game_of_life.gif`. Could you help make `nbody_oscillator.gif` and share it with me? If you are busy I'll do that on my own since this PR should be merged as soon as possible.

**Comment by KLozes** [comment]
For that last gif I could code up a simple compressible Euler equation example. Also, I think it would be easy to do wavelet AMR thanks to Taichi. The gif could look something like this https://www.youtube.com/watch?v=k72_YeNO7yU . What do you think?

**Comment by yuanming-hu** [comment]
> For that last gif I could code up a simple compressible Euler equation example. Also, I think it would be easy to do wavelet AMR thanks to Taichi. The gif could look something like this https://www.youtube.com/watch?v=k72_YeNO7yU . What do you think?

The video looks fabulous!! For this PR I think 8 gifs should be good, but in the future, we can definitely extend the gallery to 12 or 16 items. 

On one hand, if we can have a compressible Euler equation solver gif, that would be fantastic! On the other hand, how long do you expect a demo like that to be done? No rush on that, but we should try to update the README.md file as soon as possible (ideally within 48h), since the current README doesn't have too much useful information... If that demo is going to take a while or you are busy recently, I'll just make a gif out of an existing demo. (But I'm more than happy to include yours when it's ready! :-)

**Comment by KLozes** [comment]
Thanks! I can definitely have a uniform grid version done in a night or two. Wavelet AMR will take a few weeks I imagine, but will be a very cool example!

**Comment by yuanming-hu** [comment]
> Thanks! I can definitely have a uniform grid version done in a night or two. Wavelet AMR will take a few weeks I imagine, but will be a very cool example!

Sounds great! Let's leave the final slot for your demo then! A uniform grid version would be good enough. I'm sure people will be impressed. (The small scale vortices in your Youtube video was really impressive!)

**Comment by archibate** [comment]
Here we go!

nbody_oscillator

**Comment by KLozes** [comment]
I think an n-body example is a worthy addition to entice astrophysics people, but I think we could add a little more pizzazz to the algorithm and visualization. It would make a very good example of sparse computation if we replace the N^2 gravity computation with the Fast Multiple Method. FMM is basically the multigrid of n-body simulations. If we used this method could also display the octree grid as well, which could look pretty neat.

**Comment by archibate** [comment]
Cool! Octrees sounds like what taichi_sparse is to do.

**Comment by KLozes** [comment]
Here's the gif! 192 pixels doesn't really do it justice though. 

video

video

**Comment by yuanming-hu** [comment]
Thanks @KLozes, I sped up the gif a little bit, to make it more dramatic and to save people's network bandwidth :-)

@archibate We will definitely showcase the `n-body` example if you can implement a `O(N log N)` tree code algorithm!

@k-ye @archibate @xumingkuan @KLozes I'm merging this in for now so that people can see the fancy gifs in time. The current README page doesn't really have anything attractive. Please feel free to continue the discussions in https://github.com/taichi-dev/taichi/issues/740, on the **demos** or on the **list of developers**. 

Back in the year 2017, @squarefk contributed a lot while Taichi was not yet a compiler. Part of his code is still being used as our compiler infrastructure. Later he changed his email address (I guess) so github lost track of his commits. I'm including his name here to show our recognitions.

The Taichi community appreciates contributions from everybody. I'm trying my best to make sure the list of developers reflects everyone's contribution, and please feel free to share your thoughts (if any) on the current contributor list, in this thread or to my mailbox (yuanmhu ___a t___ gmail.com). 

I think we have made great progress since Jan 2020, and it has been **a really great pleasure** to work together :-) Thank you again for being part of our community!

**Comment by k-ye** [comment]
The feeling is mutual! And you should definitely write more articles like <99行代码的《冰雪奇缘》> (when you have time??) 😆

**Comment by yuanming-hu** [comment]
Right - we should definitely write more posts online to advocate Taichi given now we are more ready :-)

I'll try to write one after v0.6. Please feel free to write your own articles/tutorials.

**Comment by archibate** [comment]
hhh, 冰雪奇缘, I remember I first saw this on zhihu, and immediately attracted by the effects of snow :)

**Pull request merged**